linkerd/app/outbound/src/http/detect.rs

```diff
@@ -26,10 +28,54 @@ impl<T> Outbound<T> {
         HSvc: Clone + Send + Sync + Unpin + 'static,
         HSvc::Error: Into<Error>,
         HSvc::Future: Send,
+    {
+        let skipped = self
+            .clone()
+            .into_stack()
+            .push_on_response(svc::MapTargetLayer::new(io::EitherIo::Left))
+            .into_inner();
+        self.push_detect_http(http)
+            // When the profile marks the target as opaque, we skip HTTP
+            // detection and just use the TCP logical stack directly.
+            .push_switch(
+                |l: tcp::Logical| -> Result<_, Error> {
+                    if l.profile.borrow().opaque_protocol {
+                        Ok(svc::Either::B(l))
+                    } else {
+                        Ok(svc::Either::A(l))
+                    }
+                },
+                skipped,
+            )
+    }
+
+    pub fn push_detect_http<R, TSvc, TTgt, H, HSvc, HTgt, I>(
+        self,
+        http: H,
+    ) -> Outbound<
+        impl svc::NewService<
+                R,
+                Service = impl svc::Service<I, Response = (), Error = Error, Future = impl Send>,
+            > + Clone,
+    >
+    where
+        I: io::AsyncRead + io::AsyncWrite + io::PeerAddr + std::fmt::Debug + Send + Unpin + 'static,
+        T: svc::NewService<TTgt, Service = TSvc> + Clone + Send + 'static,
+        TSvc: svc::Service<io::EitherIo<I, io::PrefixedIo<I>>, Response = ()> + Send + 'static,
+        TSvc::Error: Into<Error>,
+        TSvc::Future: Send,
+        TTgt: From<R> + 'static,
```

Comment:
nit: It would probably be clearest (i.e. what we do elsewhere) to make `T` the target type (replacing `R`) and to use `N`/`NSvc`

Reply:
fixed in 8f9d9efdf1813afe370187c8c6990b414c8e93d3